pyro/distributions/affine_beta.py

```diff
@@ -0,0 +1,107 @@
+import torch
+from pyro.distributions.torch import Beta, TransformedDistribution
+from torch.distributions import constraints
+from torch.distributions.transforms import AffineTransform
+
+
+class AffineBeta(TransformedDistribution):
+    r"""
+    Beta distribution scaled by :attr:`scale` and shifted by :attr:`loc`::
+
+        X ~ Beta(concentration1, concentration0)
+        f(X) = loc + scale * X
+        Y = f(X) ~ AffineBeta(concentration1, concentration0, loc, scale)
+
+    :param float or torch.Tensor concentration1: 1st concentration parameter
+        (alpha) for the Beta distribution.
+    :param float or torch.Tensor concentration0: 2nd concentration parameter
+        (beta) for the Beta distribution.
+    :param float or torch.Tensor loc: location parameter.
+    :param float or torch.Tensor scale: scale parameter.
+    """
+
+    arg_constraints = {
+        "concentration1": constraints.positive,
+        "concentration0": constraints.positive,
+        "loc": constraints.real,
+        "scale": constraints.positive,
+    }
+
+    def __init__(self, concentration1, concentration0, loc, scale, validate_args=None):
+        base_dist = Beta(concentration1, concentration0)
+        super(AffineBeta, self).__init__(
+            base_dist,
+            AffineTransform(loc=loc, scale=scale),
+            validate_args=validate_args,
+        )
+
+    def expand(self, batch_shape, _instance=None):
+        new = self._get_checked_instance(AffineBeta, _instance)
+        return super(AffineBeta, self).expand(batch_shape, _instance=new)
+
+    def sample(self, sample_shape=torch.Size()):
+        """
+        Generates a sample from `Beta` distribution and applies `AffineTransform`.
+        Additionally clamps the output in order to avoid `NaN` and `Inf` values
+        in the gradients.
+        """
+        with torch.no_grad():
+            x = self.base_dist.sample(sample_shape)
+            for transform in self.transforms:
+                x = transform(x)
+            eps = torch.finfo(x.dtype).eps * self.scale
+            x = torch.min(torch.max(x, self.low + eps), self.high - eps)
+            return x
```

Comment:
I believe you can simply omit `.sample()`, since [Distribution.sample()](https://github.com/pytorch/pytorch/blob/0f412aa29392d75e33f7e37d8276be8b40b4ae73/torch/distributions/distribution.py#L123-L129) defaults to `torch.no_grad()(rsample)`.

Reply:
`.sample()` is redefined in [`TransformedDistribution`](https://github.com/pytorch/pytorch/blob/0f412aa29392d75e33f7e37d8276be8b40b4ae73/torch/distributions/transformed_distribution.py#L78-L89). Maybe I should add it back.

Reply:
ok, this is fine then.